searx/botdetection/ip_limit.py

```diff
@@ -0,0 +1,146 @@
+bot request.  This method requires a redis DB and needs a HTTP X-Forwarded-For_
+header.  To take privacy only the hash value of an IP is stored in the redis DB
+and at least for a maximum of 10 minutes.
+
+The :py:obj:`.link_token` method can be used to investigate whether a request is
+*suspicious*.  To activate the :py:obj:`.link_token` method in the
+:py:obj:`.ip_limit` method add the following to your
+``/etc/searxng/limiter.toml``:
+
+.. code:: toml
+
+   [botdetection.ip_limit]
+   link_token = true
+
+If the :py:obj:`.link_token` method is activated and a request is *suspicious*
+the request rates are reduced:
+
+- :py:obj:`BURST_MAX` -> :py:obj:`BURST_MAX_SUSPICIOUS`
+- :py:obj:`LONG_MAX` -> :py:obj:`LONG_MAX_SUSPICIOUS`
+
+To intercept bots that get their IPs from a range of IPs, there is a
+:py:obj:`SUSPICIOUS_IP_WINDOW`.  In this window the suspicious IPs are stored
+for a longer time.  IPs stored in this sliding window have a maximum of
+:py:obj:`SUSPICIOUS_IP_MAX` accesses before they are blocked.  As soon as the IP
+makes a request that is not suspicious, the sliding window for this IP is
+droped.
+
+.. _X-Forwarded-For:
+   https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For
+
+"""
+from __future__ import annotations
+from ipaddress import (
+    IPv4Network,
+    IPv6Network,
+)
+
+import flask
+import werkzeug
+from searx.tools import config
+
+from searx import redisdb
+from searx import logger
+from searx.redislib import incr_sliding_window, drop_counter
+
+from . import link_token
+from ._helpers import too_many_requests
+
+
+logger = logger.getChild('botdetection.ip_limit')
+
+BURST_WINDOW = 20
+"""Time (sec) before sliding window for *burst* requests expires."""
+
+BURST_MAX = 15
+"""Maximum requests from one IP in the :py:obj:`BURST_WINDOW`"""
+
+BURST_MAX_SUSPICIOUS = 2
+"""Maximum of suspicious requests from one IP in the :py:obj:`BURST_WINDOW`"""
+
+LONG_WINDOW = 600
+"""Time (sec) before the longer sliding window expires."""
+
+LONG_MAX = 150
+"""Maximum requests from one IP in the :py:obj:`LONG_WINDOW`"""
+
+LONG_MAX_SUSPICIOUS = 10
+"""Maximum suspicious requests from one IP in the :py:obj:`LONG_WINDOW`"""
+
+API_WONDOW = 3600
+"""Time (sec) before sliding window for API requests (format != html) expires."""
+
+API_MAX = 4
+"""Maximum requests from one IP in the :py:obj:`API_WONDOW`"""
+
+SUSPICIOUS_IP_WINDOW = 3600 * 24 * 30
+"""Time (sec) before sliding window for one suspicious IP expires."""
+
+SUSPICIOUS_IP_MAX = 3
+"""Maximum requests from one suspicious IP in the :py:obj:`SUSPICIOUS_IP_WINDOW`."""
+
+
+def filter_request(
+    network: IPv4Network | IPv6Network,
+    request: flask.Request,
+    cfg: config.Config,
+) -> werkzeug.Response | None:
+
+    # pylint: disable=too-many-return-statements
+    redis_client = redisdb.client()
+
+    if network.is_link_local and not cfg['botdetection.ip_limit.filter_link_local']:
```

Comment:
Where is the list of link local addresses?

Reply:
The object `network` is a instance of class [IPv4Network](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Network) or [IPv6Network](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv6Network). These classes are python builtin types to handle IP addresses & networks.  The documentation:
-  [is_link_local](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_link_local) --> constant definition in the code for [IPv4](https://github.com/python/cpython/blob/77d25795862f19c6e3d647b76cfb10d5ce1f149c/Lib/ipaddress.py#L1542-L1570) and [IPv6](https://github.com/python/cpython/blob/77d25795862f19c6e3d647b76cfb10d5ce1f149c/Lib/ipaddress.py#L2262-L2292)

In the configuration [searx/botdetection/limiter.toml](https://github.com/searxng/searxng/pull/2357/commits/281e36f4b7848374535d5e953050ae73423191ca#diff-437a93b109d23719837e77f803998a188751dcc7f60cf711996939653defa967) link local addresses are unfiltered by default .. in a follow-up PR we can add an additional option for [is_private](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_private)

Reply:
Shouldn't we offer the ability to add more link local networks? Some internal networks use public ranges, that's actually very common in IPv6 as there is usually no NAT involved and the "local" network is the public range of the network of computers behind the "modem" (internet connection).

Reply:
Yes I've used the word link local, but I actually meant private networks.

Maybe you should rename the variable and function names to "private_networks" if all the addresses located here are "whitelisted": https://github.com/python/cpython/blob/77d25795862f19c6e3d647b76cfb10d5ce1f149c/Lib/ipaddress.py#L2262-L2292

We can open another pull request for adding a support for more private networks.

Reply:
> Maybe you should rename the variable and function names to "private_networks" if all the addresses located here are "whitelisted":

Sorry, my link was not clear, "link local" is only:

- `IPv6Network('fe80::/10')`
- `IPv4Network('169.254.0.0/16')`

by this, the name `link_locale` is OK / I think.

> We can open another pull request for adding a support for more private networks.

:+1:

Do you see any other questionable in this PR?

Reply:
No, options for `_private_networks` and  `_reserved_networks` I would add in a follow up PR, I just linked them all to show, that there other properties to check [network.is_private](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_private) and [netwirk.is_reserved](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_reserved) ...

Reply:
Ok then we can tackle this issue in another PR.